src/coordination/coordinator_state_machine.cpp

```diff
@@ -12,14 +12,136 @@
+  return nlohmann::json{
+      {kCoordClusterState, snapshot_ctx.cluster_state_.SerializeJson().dump()},
+      {kLastLogTerm, snapshot_ctx.snapshot_->get_last_log_term()},
+      {kLastLogIdx, snapshot_ctx.snapshot_->get_last_log_idx()},
+      {kSize, snapshot_ctx.snapshot_->size()},
+      {kLastConfig, memgraph::coordination::SerializeClusterConfig(*snapshot_ctx.snapshot_->get_last_config()).dump()},
+      {kType, FromSnapshotType(snapshot_ctx.snapshot_->get_type())}};
+}
+
 }  // namespace
 
 namespace memgraph::coordination {
 
-CoordinatorStateMachine::CoordinatorStateMachine(LoggerWrapper logger) : logger_(logger) {}
+CoordinatorStateMachine::CoordinatorStateMachine(std::shared_ptr<kvstore::KVStore> durability, LoggerWrapper logger)
+    : logger_(logger), durability_store_(std::move(durability)) {
+  if (!durability_store_) {
+    logger_.Log(nuraft_log_level::INFO, "Coordinator state machine stores snapshots only in memory from now on.");
+    return;
+  }
+
+  logger_.Log(nuraft_log_level::INFO, "Restoring coordinator state machine with durability.");
+
+  int version{0};
+  auto maybe_version = durability_store_->Get(kSnapshotVersion);
+  if (maybe_version.has_value()) {
+    version = std::stoi(maybe_version.value());
+  } else {
+    logger_.Log(
+        nuraft_log_level::TRACE,
+        fmt::format(
+            "Assuming first start of coordinator state machine using durability as version is missing, storing current "
+            "active version {}.",
+            kActiveVersion));
+    MG_ASSERT(durability_store_->Put(kSnapshotVersion, std::to_string(kActiveVersion)),
+              "Failed to store version to disk");
+    version = 1;
+  }
+
+  MG_ASSERT(version <= kActiveVersion && version > 0, "Unsupported version of log store with durability");
+
+  for (auto kv_store_snapshot_it = durability_store_->begin(std::string{kSnapshotIdPrefix});
+       kv_store_snapshot_it != durability_store_->end(std::string{kSnapshotIdPrefix}); ++kv_store_snapshot_it) {
+    auto const &[snapshot_key_id, snapshot_key_value] = *kv_store_snapshot_it;
+    try {
+      auto parsed_snapshot_id =
+          std::stoul(std::regex_replace(snapshot_key_id, std::regex{kSnapshotIdPrefix.data()}, ""));
+      last_committed_idx_ = std::max(last_committed_idx_.load(), parsed_snapshot_id);
+      snapshots_[parsed_snapshot_id] = DeserializeSnapshotCtxFromDisk(snapshot_key_id, *durability_store_);
+      MG_ASSERT(parsed_snapshot_id == snapshots_[parsed_snapshot_id]->snapshot_->get_last_log_idx(),
+                "Parsed snapshot id {} does not match last log index {}", parsed_snapshot_id,
+                snapshots_[parsed_snapshot_id]->snapshot_->get_last_log_idx());
+    } catch (std::exception &e) {
+      LOG_FATAL("Failed to deserialize snapshot with id: {}. Error: {}", snapshot_key_id, e.what());
+    }
+    logger_.Log(nuraft_log_level::TRACE, fmt::format("Deserialized snapshot with id: {}", snapshot_key_id));
```

Comment:
Can you explain intention here with regexes and using RocksDB prefixes capabilities -> why, can we make it simpler, these prefix things in RocksDB aren't trivial

Reply:
Let's jump for this one on call

Reply:
resolved